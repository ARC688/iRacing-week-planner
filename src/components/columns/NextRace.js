import React, { Component, PropTypes } from 'react';

export default class NextRace extends Component {
  static propTypes = {
    race: PropTypes.object.isRequired
  }

  render() {
    const { race } = this.props;

    if (race.nextTime === null) {
      return <td>No time data</td>;
    }

    return <td>{race.nextTime.local().format('ddd h:mma')}</td>;
  }
}
