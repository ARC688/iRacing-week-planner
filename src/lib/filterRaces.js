import {clone, intersection} from 'lodash';

export default function({
  races, filters, ownedTracks, ownedCars, favouriteSeries, favouriteCars, favouriteTracks
}) {
  let filteredRaces = clone(races);

  filteredRaces = filteredRaces.filter((race) => {
    return filters.type.indexOf(race.type) !== -1;
  });

  filteredRaces = filteredRaces.filter((race) => {
    return filters.licence.indexOf(race.licenceClass) !== -1;
  });

  filteredRaces = filteredRaces.filter((race) => {
    return filters.fixed.indexOf(race.fixed) !== -1;
  });

  filteredRaces = filteredRaces.filter((race) => {
    return filters.official.indexOf(race.official) !== -1;
  });

  if (filters.ownedTracks) {
    filteredRaces = filteredRaces.filter((race) => {
      return ownedTracks.indexOf(race.trackId) !== -1;
    });
  }

  if (filters.ownedCars) {
    filteredRaces = filteredRaces.filter((race) => {
      return intersection(ownedCars, race.carIds).length !== 0;
    });
  }

  if (filters.favouriteSeries) {
    filteredRaces = filteredRaces.filter((race) => {
      return favouriteSeries.indexOf(race.seriesId) !== -1;
    });
  }

  if (filters.favouriteCarsOnly) {
    filteredRaces = filteredRaces.filter((race) => {
      return intersection(favouriteCars, race.carIds).length !== 0;
    });
  }

  if (filters.favouriteTracksOnly) {
    filteredRaces = filteredRaces.filter((race) => {
      return favouriteTracks.indexOf(race.trackId) !== -1;
    });
  }

  return filteredRaces;
}
