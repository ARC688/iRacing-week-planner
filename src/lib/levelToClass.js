// @flow

const levels: Array<string> = ['R', 'D', 'C', 'B', 'A', 'P', 'P'];

export default function levelToClass(level: number, effective: boolean = false): string {
  return effective ? effectiveLevelToClass(level) : realLevelToClass(level);
}

function effectiveLevelToClass(level: number): string {
  const index = Math.floor((level) / 4);

  return levels[index];
}

function realLevelToClass(level: number): string {
  const index = Math.floor((level - 1) / 4);

  return levels[index];
}

export function levelToClassNumber(level: number): number {
  return Math.floor((level) / 4);
}
