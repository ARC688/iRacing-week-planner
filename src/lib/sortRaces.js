import { cloneDeep } from 'lodash';

export default (rules, unordered) => {
  const races = cloneDeep(unordered);
  races.sort((a, b) => {
    for (const rule of rules) {
      if (a[rule.key] === b[rule.key]) {
        continue;
      }
      if (rule.order === 'asc') {
        return a[rule.key] < b[rule.key] ? -1 : 1;
      }
      if (rule.order === 'desc') {
        return a[rule.key] > b[rule.key] ? -1 : 1;
      }
    }
    return 0;
  });
  return races;
};
