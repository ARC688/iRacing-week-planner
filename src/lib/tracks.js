import allTracks from '../data/tracks.json';
import {uniqBy, sortBy} from 'lodash';

const fixText = (text) => (decodeURIComponent(text).replace(/\+/g, ' ').trim());

const tracks = sortBy(uniqBy(sortBy(allTracks, 'priority'), track => track.pkgid), track => track.name);

export default tracks.map((track) => {
  return {
    id: track.pkgid,
    name: track.skuname ? fixText(track.skuname) : fixText(track.name),
    hasOval: allTracks.filter((t) => t.pkgid === track.pkgid && t.catid === 1).length > 0,
    hasRoad: allTracks.filter((t) => t.pkgid === track.pkgid && t.catid === 2).length > 0,
    primaryType: track.catid === 2 ? 'road' : 'oval',
    default: track.freeWithSubscription === true
  };
});
