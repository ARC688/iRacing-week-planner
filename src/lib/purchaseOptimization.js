import series from '../data/season.json';
import tracks from '../data/tracks';

export default function({
  ownedTracks, favouriteSeries,
}) {
  const currentSeries = favouriteSeries.length === 0 ? series : series.filter(serie => favouriteSeries.includes(serie.seriesid))
  const allTrackPkgIds = currentSeries
    .flatMap(serie => serie.tracks.map(track => track.pkgid))
    .filter(pkgId => !ownedTracks.includes(pkgId))
  const countById = Object.values(allTrackPkgIds.reduce((map, trackPkgId) => {
    const originalTrack = tracks.find(track => track.pkgid === trackPkgId)
    const fromSeries = currentSeries.filter(serie => serie.tracks.filter(serieTrack => serieTrack.pkgid === trackPkgId).length > 0).map(serie => serie.seriesname)
    map[trackPkgId] = map[trackPkgId] || {
      id: originalTrack.id,
      name: originalTrack.name,
      series: fromSeries,
      count: 0
    }
    map[trackPkgId].count += 1
    return map
  }, {}))
  // Fill series
  // Sort series by name

  console.log(countById)
  return countById
    .filter(item => item.count > 1)
    .sort((a, b) => b.count - a.count); // Sort by count DESC / name ASC
}

