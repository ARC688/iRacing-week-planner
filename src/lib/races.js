import season from '../data/season.json';
import levelToClass, {levelToClassNumber} from './levelToClass';
import raceTimesArray from '../data/raceTimes';
import moment, {duration} from 'moment';

import tracks from '../data/tracks.json';

const tracksById = tracks.reduce((tracksObj, track) => {
  return { ...tracksObj, [track.id]: track };
}, {});

const raceTimesById = raceTimesArray.reduce((races, race) => {
  return { ...races, [race.seriesId]: race };
}, {});

const now = moment().utc();
const startOfWeek = now.clone().subtract(1, 'days').startOf('isoWeek').add(1, 'days');

function getNextRaceFromRecur(everyTime, offset) {
  const nextDate = startOfWeek.clone().add(offset);
  const endDate = startOfWeek.clone().add({ weeks: 1, days: 1});

  while (nextDate.isBefore(now) && nextDate.isBefore(endDate)) {
    nextDate.add(everyTime);
  }

  if (nextDate.isBefore(now)) {
    return null;
  }

  return nextDate;
}

function getNextRaceSetTimes(setTimes) {
  for (const time of setTimes) {
    const date = startOfWeek.clone().add(time);
    if (date.isAfter(now)) {
      return date;
    }
  }

  const firstNextWeek = startOfWeek.clone().add(1, 'weeks').add(setTimes[0]);

  if (firstNextWeek.isAfter(now)) {
    return firstNextWeek;
  }

  return null;
}

const fixText = (text) => (decodeURIComponent(text).replace(/\+/g, ' ').trim());

export default season.reduce((carry, series) => {
  const raceTimes = raceTimesById[series.seriesid] || {};

  const seriesName = fixText(series.seriesname);
  const seriesStart = moment(series.start, 'x').utc().startOf('day');

  if (raceTimes.weekStartOffset) {
    seriesStart.add(raceTimes.weekStartOffset);
  }

  const seriesEnd = moment(series.end, 'x').utc().startOf('isoWeek').add({days: 1});

  if (raceTimes.weekEndOffset) {
    seriesEnd.add(raceTimes.weekEndOffset);
  }

  const offWeeks = raceTimes.offWeeks || [];
  const raceWeekLength = Math.round(moment(seriesEnd).diff(seriesStart) / (series.tracks.length + offWeeks.length));

  const allRaceWeeks = series.tracks.map(track => track.raceweek)
    .concat(offWeeks.map(offWeek => offWeek - 1));

  allRaceWeeks.sort((a, b) => a - b);

  return carry.concat(series.tracks.map((track) => {
    const trackName = track.config ? `${track.name} - ${track.config}` : track.name;
    let nextTime = null;

    const realRaceWeek = allRaceWeeks.indexOf(track.raceweek);

    if (raceTimes.everyTime) {
      nextTime = getNextRaceFromRecur(raceTimes.everyTime, raceTimes.offset);
    }

    if (raceTimes.setTimes) {
      nextTime = getNextRaceSetTimes(raceTimes.setTimes);
    }

    const startTime = moment(seriesStart).add(raceWeekLength * realRaceWeek, 'ms').startOf('day').utc();
    const weekLength = duration(raceWeekLength);

    let type = series.catid === 1 ? 'Oval' : 'Road';

    if (tracksById[track.id] !== undefined && tracksById[track.id].isDirt) {
      type = 'Dirt';
    }

    return {
      series: seriesName,
      seriesId: series.seriesid,
      track: fixText(trackName),
      trackId: track.pkgid,
      week: track.raceweek,
      startTime,
      weekLength,
      endTime: moment(startTime).add(weekLength),
      official: series.isOfficial,
      licenceLevel: series.minlicenselevel,
      licenceClassNumber: levelToClassNumber(series.minlicenselevel),
      licenceClass: levelToClass(series.minlicenselevel, true),
      type,
      fixed: series.isFixedSetup,
      carClasses: series.carclasses.map((carClass) => fixText(carClass.shortname)),
      carIds: series.cars.map((car) => car.sku),
      raceTimes,
      nextTime,
      seriesStart,
      seriesEnd,
      seasonId: series.seasonid
    };
  }));
}, []);
