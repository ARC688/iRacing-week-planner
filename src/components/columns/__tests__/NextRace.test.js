import MockDate from 'mockdate';
import moment from 'moment';
import React from 'react';
import { shallow } from 'enzyme';
import { describe, test } from '@jest/globals';
import { getNextRace } from '../../../lib/races';

import NextRace from '../NextRace';

jest.mock(
  '../../../lib/races',
  () => ({
    getNextRace: jest.fn(),
  }),
);

describe('components/columns/NextRace', () => {
  test('renders correctly', () => {
    MockDate.set('2020-09-23T13:30:00.000Z');
    getNextRace.mockReturnValue(moment('2020-09-23T13:45:00.000Z'));
    const component = shallow(<NextRace race={{}} />);

    expect(component).toMatchSnapshot();
    MockDate.reset();
  });
});
