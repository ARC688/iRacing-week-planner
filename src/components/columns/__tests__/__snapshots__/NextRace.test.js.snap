// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/columns/NextRace renders correctly 1`] = `
<td>
  <div>
    Wed 14:45
  </div>
</td>
`;
